### PR TITLE
Replace Makefile build back end with SConstruct

### PR DESCRIPTION
**Pull request by jtcooper10**
Replacing the GNUMake-based build system with [SCons](https://scons.org/), which is added as a Python dependency.

Still, only `g++` is supported as the C++ compiler, which requires `mingw` and GNU tools on Windows and MacOS. In the future, using SCons rather than Make will allow us to support a variety of platform compilers more easily.

---

# Changes
- Added `SConstruct` file and several `SConscript` files
- Modified the `Make` class to use `-m SCons` instead of `make`
- Removed unneeded Makefile and env scripts

### CODE REVIEW

File: gillespy2/solvers/cpp/build/build_engine.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def __init__(self, debug: bool = False, output_dir: str = None):
             self.makefile = self.cpp_dir.joinpath("Makefile.win32")
         else:
             self.makefile = self.cpp_dir.joinpath("Makefile")
```
Comment (BryanRumsey) [inline gillespy2/solvers/cpp/build/build_engine.py:41]: Remove lines 38 - 41.
Line 42 re-assigns self.makefile.

**Review by BryanRumsey — CHANGES_REQUESTED** [review]
Need to add copyright block to all new files.

**Review by briandrawert — APPROVED** [review]

**Comment by mdip226** [comment]
Good luck reading this. It's definitely having a permission error creating a tempfile. (Line 9) I think that makes sense, because it looks like it's trying to make it in C:/WINDOWS which I assume is not allowed.
[out.txt](https://github.com/StochSS/GillesPy2/files/9795986/out.txt)

**Comment by mdip226** [comment]
Here's the important part:
`"name": "SimulationError",
	"message": "<class 'gillespy2.solvers.cpp.ssa_c_solver.SSACSolver'> is not a valid solver.  Reason Given: Error encountered during execution of Makefile target: 'ssa'.\nReturn code: 2- stdout: scons: Entering directory `C:\\Users\\mdip2\\AppData\\Local\\Temp\\gillespy2_build_695gskgr'\r\nscons: Reading SConscript files ...\r\nscons: done reading SConscript files.\r\nscons: Building targets ...\r\ng++ -o gillespy2_obj\\arg_parser.o -c -std=c++14 -IC:\\Users\\mdip2\\Projects\\GillesPy2\\gillespy2\\solvers\\cpp\\c_base -Igillespy2_template -IC:\\Users\\mdip2\\Projects\\GillesPy2\\gillespy2\\solvers\\cpp\\c_base\\Sundials\\include -IC:\\Users\\mdip2\\Projects\\GillesPy2\\gillespy2\\solvers\\cpp\\c_base\\Tau C:\\Users\\mdip2\\Projects\\GillesPy2\\gillespy2\\solvers\\cpp\\c_base\\arg_parser.cpp\r\nscons: building terminated because of errors.\r\n\n- stderr: Cannot create temporary file in C:\\WINDOWS\\: Permission denied\r\nscons: *** [gillespy2_obj\\arg_parser.o] Error 3\r\n\n- make_cmd: ['C:\\\\Users\\\\mdip2\\\\AppData\\\\Local\\\\Microsoft\\\\WindowsApps\\\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\\\python.exe', '-m', 'SCons', '-CC:\\\\Users\\\\mdip2\\\\AppData\\\\Local\\\\Temp\\\\gillespy2_build_695gskgr', '-fC:\\\\Users\\\\mdip2\\\\Projects\\\\GillesPy2\\\\gillespy2\\\\solvers\\\\cpp\\\\build\\\\SConstruct', 'CBASE_DIR=C:\\\\Users\\\\mdip2\\\\Projects\\\\GillesPy2\\\\gillespy2\\\\solvers\\\\cpp\\\\c_base', 'OBJ_DIR=C:\\\\Users\\\\mdip2\\\\AppData\\\\Local\\\\Temp\\\\gillespy2_build_695gskgr\\\\gillespy2_obj', 'TEMPLATE_DIR=C:\\\\Users\\\\mdip2\\\\AppData\\\\Local\\\\Temp\\\\gillespy2_build_695gskgr\\\\gillespy2_template', 'SOLVER=ssa', 'OUTPUT_FILE=C:\\\\Users\\\\mdip2\\\\AppData\\\\Local\\\\Temp\\\\gillespy2_build_695gskgr\\\\GillesPy2_Simulation.exe']\n- os.listdir(C:\\Users\\mdip2\\Projects\\GillesPy2\\gillespy2\\solvers\\cpp\\c_base\\template): ['template.cpp', 'template.h', 'template_defaults.h', 'template_definitions.h', 'template_opts.h', 'template_params.h']\n."`

**Review by mdip226 — CHANGES_REQUESTED** [review]
See my comments

**Comment by mdip226** [comment]
This was when I was trying to run StartHere.ipynb (not compute) on windows,
by bad I should have included more info. Does the scons python package
require the windows executable scons.exe to be installed?

**Comment by mdip226** [comment]
Its the error message produced by Jupyter, opened in a text editor cuz the
error message was big

**Comment by mdip226** [comment]
It was a hasty attempt that involved
1. cloning Gillespie repo
2. Installing mingw along with g++, make (I actually had a little trouble locating the correct make and I'm still not 100% sure its the correct one but I think it is. Both commands seem to run from cmd but I never tested them farther than knowing they are in the path.
3. pip install scons
4. Running starthere.ipynb in vscode

**Review by BryanRumsey — CHANGES_REQUESTED** [review]
Integration with StochSS fails on staging and local with Apple M1 chip.

## Error from Tests
```
ERROR: test_ode_c_solver (test_gillespy2.TestGillesPy2Dependency) (model='Brusselator')
Check if the test_models run with the ODECSolver.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/gillespy2/core/model.py", line 1282, in run
    return solver.run(t=t, increment=increment, timeout=timeout, **solver_args)
  File "/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/ode_c_solver.py", line 177, in run
    sim_status = self._run(sim_exec, args, decoder, timeout, display_args)
  File "/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_solver.py", line 217, in _run
    with subprocess.Popen(sim_args, stdout=subprocess.PIPE, **platform_args) as simulation:
  File "/opt/conda/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/conda/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/gillespy2_build_t5ngffyr/GillesPy2_Simulation.out'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/stochss/stochss/tests/test_gillespy2.py", line 109, in test_ode_c_solver
    test_model.run(solver=test_solver)
  File "/opt/conda/lib/python3.8/site-packages/gillespy2/core/model.py", line 1284, in run
    raise SimulationError(
gillespy2.core.gillespyError.SimulationError: argument 'solver=<gillespy2.solvers.cpp.ode_c_solver.ODECSolver object at 0x7f5c92f72250>' to run() failed.  Reason Given: [Errno 2] No such file or directory: '/tmp/gillespy2_build_t5ngffyr/GillesPy2_Simulation.out'
```

## Error from Job on staging
```
2022-11-16 04:51:15,882 - GillesPy2 - ERROR - Job errors: argument 'solver=' to run() failed. Reason Given: [Errno 2] No such file or directory: '/tmp/gillespy2_build__cbm2cey/GillesPy2_Simulation.out'
Traceback (most recent call last):
File "/opt/conda/lib/python3.8/site-packages/gillespy2/core/model.py", line 1282, in run
return solver.run(t=t, increment=increment, timeout=timeout, **solver_args)
File "/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/ssa_c_solver.py", line 166, in run
sim_status = self._run(sim_exec, args, decoder, timeout, display_args)
File "/opt/conda/lib/python3.8/site-packages/gillespy2/solvers/cpp/c_solver.py", line 217, in _run
with subprocess.Popen(sim_args, stdout=subprocess.PIPE, **platform_args) as simulation:
File "/opt/conda/lib/python3.8/subprocess.py", line 854, in __init__
self._execute_child(args, executable, preexec_fn, close_fds,
File "/opt/conda/lib/python3.8/subprocess.py", line 1702, in _execute_child
raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/gillespy2_build__cbm2cey/GillesPy2_Simulation.out'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/stochss/stochss/handlers/util/scripts/start_job.py", line 118, in
job.run(verbose=args.verbose)
File "/stochss/stochss/handlers/util/ensemble_simulation.py", line 125, in run
results = self.g_model.run(number_of_trajectories=1 if is_ode else 100)
File "/opt/conda/lib/python3.8/site-packages/gillespy2/core/model.py", line 1284, in run
raise SimulationError(
gillespy2.core.gillespyError.SimulationError: argument 'solver=' to run() failed. Reason Given: [Errno 2] No such file or directory: '/tmp/gillespy2_build__cbm2cey/GillesPy2_Simulation.out'
```

**Pull request merged**